src/replication/log_generator.cpp

```diff
@@ -37,311 +37,508 @@ namespace cubreplication
     LOG_TDES *tdes = LOG_FIND_TDES (tran_index);
 
-    set_tran_repl_info (tdes->mvccinfo.id, state);
-    pack_stream_entry ();
-  }
+    MVCCID mvccid = logtb_find_current_mvccid (&cubthread::get_entry ());
+    set_tran_repl_info (mvccid, stream_entry_header::ACTIVE);
+
+    /* Write objects in stream and then destroy them. */
+    local_stream_entry.pack ();
+    local_stream_entry.reset ();
+  }  
 
   void
-  log_generator::on_transaction_commit (void)
+    log_generator::on_transaction_abort(void)
   {
-    on_transaction_finish (stream_entry_header::TRAN_STATE::COMMITTED);
+    on_transaction_finish (stream_entry_header::TRAN_STATE::ABORTED);
   }
 
-  void
-  log_generator::on_transaction_abort (void)
+  void log_generator::on_sysop_abort (LOG_LSA &start_lsa)
   {
-    on_transaction_finish (stream_entry_header::TRAN_STATE::ABORTED);
+    cubthread::entry *thread_p = &cubthread::get_entry();
+    cubreplication::stream_entry *stream_entry = logtb_get_tdes (thread_p)->replication_log_generator.get_stream_entry ();
+    int count_entries = (int) stream_entry->count_entries ();
+
+    if (count_entries == 0)
+    {
+      return;
+    }
+
+    stream_entry->destroy_objects_after_lsa (start_lsa);    
   }
 
   void
-  log_generator::clear_transaction (void)
+    log_generator::clear_transaction(void)
   {
-    if (is_replication_disabled ())
-      {
-	return;
-      }
+    if (is_replication_disabled())
+    {
+      return;
+    }
 
     m_is_row_replication_disabled = false;
   }
 
   bool
-  log_generator::is_replication_disabled ()
+    log_generator::is_replication_disabled()
   {
 #if defined (SERVER_MODE)
-    return !log_does_allow_replication ();
+    return !log_does_allow_replication();
 #else
     return true;
 #endif
   }
 
   bool
-  log_generator::is_row_replication_disabled ()
+    log_generator::is_row_replication_disabled()
   {
-    return is_replication_disabled () || m_is_row_replication_disabled;
+    return is_replication_disabled() || m_is_row_replication_disabled;
   }
 
   void
-  log_generator::set_row_replication_disabled (bool disable_if_true)
+    log_generator::set_row_replication_disabled(bool disable_if_true)
   {
     m_is_row_replication_disabled = disable_if_true;
   }
+  
+#if !defined(NDEBUG) && defined (SERVER_MODE)
+  int log_generator::abort_sysop_and_simulate_apply_repl_on_master (LOG_LSA &filter_replication_lsa)
+  {
+    /* This function is used for debug purpose only. */
+    int err_code = NO_ERROR;
+    replication_object *repl_obj;
+    cubthread::entry *thread_p = &cubthread::get_entry();       
+    LOG_TDES * tdes = logtb_get_tdes (&cubthread::get_entry());
+    cubreplication::stream_entry *stream_entry = tdes->replication_log_generator.get_stream_entry ();
+
+    assert(stream_entry->count_entries() > 0);
+    
+    /* Save the replication objects, before abort. */
+    std::vector <cubreplication::replication_object *> repl_objects_after_lsa;
```

Comment:
I propose to get rid of move_replication_objects_after_lsa (with vector <replication_object> argument), and keep only the variant with stream_entry argument.

We need an "apply" method on the  cubreplication::stream_entry which applies all replication_objects.
Please see issue http://jira.cubrid.org/browse/CBRD-22796.